### PR TITLE
🐛 Fix an error when trying to save a process that contains a UserTask with no FormFields 

### LINKED ISSUE #771: 🐛Saving Diagram fails on PE when User Task has no Element
#### Describe your issue

If you create a diagram with a single user task and don't give it any elements, you can save it on your filesystem but not on your PE.
The Problem here is that the user doesn't get informed about this  error before saving.

However there is no request at the PE so this is not a Backend error but an error in the saveDiagram-Function of the process-def-detail component.

See this Screenshots for more details:

The diagram:
bildschirmfoto 2018-08-07 um 11 02 16

Error in Toastr:
bildschirmfoto 2018-08-07 um 11 00 46

Error in console:
bildschirmfoto 2018-08-07 um 11 02 02

Fun fact:
If you add a Formfield to the user task and delete it right away you can now save the diagram.

#### Possible solution

Look for errors in validation that gets called before saving the diagram in process-def-detail view.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by robpal94**
:tada: **Thanks for submitting this Pull Request**

Please provide a list including your changes.

To help us to create a meaningful Changelog, we agreed 
to use an emoji-based commit-style. Use the [Emojis](#emojis)
we provided below.

Fill out the following, provide reference to issues
and this PR- we will use this as merge commit message-body
and the title of this PR as commit-headline:

---

**Changes:**

1. Fix a bug in the `_dropInvalidFormData()` Method from the _Processdef-detail_ Component. 
This bug leads to a _Cannot read property length of undefined_ Error when trying to save a Diagram that contains a UserTask with no defined FormFields.

2. Add a comment to describe a kinda strange behavior.

**Issues:**

Closes #771 

---

See the [examples](#example) for inspiration.

## How can others test the changes?
* Create a Process with a UserTask but no form fields (**don't touch the form field section!**)
* Deploy this Process to the ProcessEngine
* Try to save this Process

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: src/contracts/bpmnmodeler/IElementRegistry.ts [hunk 1]
```diff
@@ -1,6 +1,89 @@
...
+   *
+   * @param filter Filter for selecting the element
+   */
+  get(filter: string | SVGElement): IShape;
+
+  /**
+   * Returns all elements that match a given filter function.
+   *
+   * @param filterMethod Filter function
+   * @return The Elements that matches the fiter function.
```
Comment (Vyperus) [inline src/contracts/bpmnmodeler/IElementRegistry.ts]: Typo: filter.

File: src/modules/processdef-detail/processdef-detail.ts [hunk 2]
```diff
@@ -412,7 +413,16 @@ export class ProcessDefDetail {
           const extensions: IExtensionElement = businessObj.extensionElements;
 
           extensions.values = extensions.values.filter((value: IFormElement) => {
-            const keepThisValue: boolean = value.$type !== 'camunda:FormData' || value.fields.length > 0;
+
+            const typeIsCamundaFormData: boolean = value.$type !== 'camunda:FormData';
+            const elementContainFields: boolean = ((): boolean => {
```
Comment (Vyperus) [inline src/modules/processdef-detail/processdef-detail.ts]: Typo: elementContainsFields

**Comment by Vyperus** [comment]
If there was a bug in the processdef-detail component please double check if this error occurs in the diagram-detail component too.

File: src/modules/processdef-detail/processdef-detail.ts [hunk 3]
```diff
@@ -412,7 +413,16 @@ export class ProcessDefDetail {
...
 
           extensions.values = extensions.values.filter((value: IFormElement) => {
-            const keepThisValue: boolean = value.$type !== 'camunda:FormData' || value.fields.length > 0;
+
+            const typeIsCamundaFormData: boolean = value.$type !== 'camunda:FormData';
+            const elementContainsFields: boolean = ((): boolean => {
+              const valueFieldsAreDefined: boolean = value.fields !== undefined;
+              return (valueFieldsAreDefined)
+                      ? (value.fields.length > 0)
+                      : (false);
```
Comment (LeoTT) [inline src/modules/processdef-detail/processdef-detail.ts]: I would prefer this:
```ts
const valueFieldsAreDefined: boolean = (value.fields !== undefined) && (value.fields.length > 0);
return valueFieldsAreDefined;
```

File: src/modules/processdef-detail/processdef-detail.ts [hunk 4]
```diff
@@ -412,7 +413,16 @@ export class ProcessDefDetail {
           const extensions: IExtensionElement = businessObj.extensionElements;
 
           extensions.values = extensions.values.filter((value: IFormElement) => {
-            const keepThisValue: boolean = value.$type !== 'camunda:FormData' || value.fields.length > 0;
+
+            const typeIsCamundaFormData: boolean = value.$type !== 'camunda:FormData';
```
Comment (LeoTT) [inline src/modules/processdef-detail/processdef-detail.ts]: The variable says `typeIsCamundaFormData`, but then it says `!== 'camunda:FormData'`. Please doublecheck.

File: src/modules/processdef-detail/processdef-detail.ts [hunk 5]
```diff
@@ -412,7 +413,16 @@ export class ProcessDefDetail {
...
+            const elementContainsFields: boolean = ((): boolean => {
+              const valueFieldsAreDefined: boolean = value.fields !== undefined;
+              return (valueFieldsAreDefined)
+                      ? (value.fields.length > 0)
+                      : (false);
+            })();
+
```
Comment (LeoTT) [inline src/modules/processdef-detail/processdef-detail.ts]: I would also drop the arrow function. Don't see the benefit, it just adds a scope and complexity imo.
```ts
            const elementContainsFields: boolean = ((): boolean => {
              const valueFieldsAreDefined: boolean = value.fields !== undefined;
              return (valueFieldsAreDefined)
                      ? (value.fields.length > 0)
                      : (false);
             }
             const keepThisValue: boolean = typeIsCamundaFormData || elementContainsFields;
```
Without arrow function:
```ts
const valueFieldsAreDefined: boolean = (value.fields !== undefined) && (value.fields.length > 0);
const keepThisValue: boolean = typeIsCamundaFormData || valueFieldsAreDefined;
```

**Comment by robpal94** [comment]
@Vyperus 
> If there was a bug in the processdef-detail component please double check if this error occurs in the diagram-detail component too.

No, because the bug was in the `_validateXML()` Method which is not present in the _diagram-detail_ Module at all.

**Pull request merged**